internal/controllers/clusterextension_controller.go

```diff
@@ -202,21 +178,133 @@ func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alp
+		if err != nil {
+			return nil, fmt.Errorf("invalid version range %q: %w", versionRange, err)
+		}
+		predicates = append(predicates, catalogfilter.InMastermindsSemverRange(vr))
+	}
+
+	if ext.Spec.UpgradeConstraintPolicy != ocv1alpha1.UpgradeConstraintPolicyIgnore && installedBundle != nil {
+		upgradePredicate, err := upgradePredicate(installedBundle)
+		if err != nil {
+			return nil, err
+		}
+
+		predicates = append(predicates, upgradePredicate)
+	}
+
+	resultSet := catalogfilter.Filter(allBundles, catalogfilter.And(predicates...))
+
+	var upgradeErrorPrefix string
+	if installedBundle != nil {
+		installedBundleVersion, err := installedBundle.Version()
+		if err != nil {
+			return nil, err
+		}
+		upgradeErrorPrefix = fmt.Sprintf("error upgrading from currently installed version %q: ", installedBundleVersion.String())
+	}
+	if len(resultSet) == 0 {
+		if versionRange != "" && channelName != "" {
+			return nil, fmt.Errorf("%sno package %q matching version %q found in channel %q", upgradeErrorPrefix, packageName, versionRange, channelName)
+		}
+		if versionRange != "" {
+			return nil, fmt.Errorf("%sno package %q matching version %q found", upgradeErrorPrefix, packageName, versionRange)
+		}
+		if channelName != "" {
+			return nil, fmt.Errorf("%sno package %q found in channel %q", upgradeErrorPrefix, packageName, channelName)
+		}
+		return nil, fmt.Errorf("%sno package %q found", upgradeErrorPrefix, packageName)
+	}
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByVersion(resultSet[i], resultSet[j])
+	})
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByDeprecated(resultSet[i], resultSet[j])
+	})
+
+	return resultSet[0], nil
+}
+
+func (r *ClusterExtensionReconciler) installedBundle(ctx context.Context, allBundles []*catalogmetadata.Bundle, ext *ocv1alpha1.ClusterExtension) (*catalogmetadata.Bundle, error) {
+	bd := &rukpakv1alpha2.BundleDeployment{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: ext.GetName()}, bd)
+	if client.IgnoreNotFound(err) != nil {
+		return nil, err
+	}
+
+	if bd.Spec.Source.Image == nil || bd.Spec.Source.Image.Ref == "" {
+		// Bundle not yet installed
+		return nil, nil
+	}
+
+	bundleImage := bd.Spec.Source.Image.Ref
+	// find corresponding bundle for the installed content
+	resultSet := catalogfilter.Filter(allBundles, catalogfilter.And(
+		catalogfilter.WithPackageName(ext.Spec.PackageName),
+		catalogfilter.WithBundleImage(bundleImage),
+	))
+	if len(resultSet) == 0 {
+		return nil, fmt.Errorf("bundle with image %q for package %q not found in available catalogs but is currently installed via BundleDeployment %q", bundleImage, ext.Spec.PackageName, bd.Name)
+	}
+
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByVersion(resultSet[i], resultSet[j])
+	})
+
+	return resultSet[0], nil
+}
+
+func upgradePredicate(installedBundle *catalogmetadata.Bundle) (catalogfilter.Predicate[catalogmetadata.Bundle], error) {
```

Comment:
Can probably this helper into `internal/controllers/successors.go` so it can be re-used in `Extension`.